src/sink/with.rs

```diff
@@ -0,0 +1,117 @@
+use core::mem;
+use core::marker::PhantomData;
+
+use {IntoFuture, Future, Poll, Async};
+use sink::{Sink, StartSend, AsyncSink};
+
+/// Sink for the `Sink::with` combinator, chaining a computation to run *prior*
+/// to pushing a value into the underlying sink.
+#[must_use = "sinks do nothing unless polled"]
+pub struct With<S, U, F, Fut>
+    where S: Sink,
+          F: FnMut(U) -> Fut,
+          Fut: IntoFuture,
+{
+    sink: S,
+    f: F,
+    state: State<Fut::Future, S::SinkItem>,
+    need_close: bool,
+    _phantom: PhantomData<fn(U)>,
+}
+
+enum State<Fut, T> {
+    Empty,
+    Process(Fut),
+    Buffered(T),
+}
+
+impl<Fut, T> State<Fut, T> {
+    fn is_empty(&self) -> bool {
+        if let State::Empty = *self {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+pub fn new<S, U, F, Fut>(sink: S, f: F) -> With<S, U, F, Fut>
+    where S: Sink,
+          F: FnMut(U) -> Fut,
+          Fut: IntoFuture<Item = S::SinkItem, Error = S::SinkError>,
+{
+    With {
+        state: State::Empty,
+        sink: sink,
+        f: f,
+        need_close: false,
+        _phantom: PhantomData,
+    }
+}
+
+impl<S, U, F, Fut> With<S, U, F, Fut>
+    where S: Sink,
+          F: FnMut(U) -> Fut,
+          Fut: IntoFuture<Item = S::SinkItem, Error = S::SinkError>,
+{
+    fn poll(&mut self) -> Poll<(), S::SinkError> {
+        loop {
+            match mem::replace(&mut self.state, State::Empty) {
+                State::Empty => break,
+                State::Process(mut fut) => {
```

Comment:
If we want to reduce moves this could perhaps be:

```rust
let item = match self.state {
    State::Empty => break,
    State::Process(ref mut f) => try_ready!(f.poll()),
    State::Buffered(_) => /* do the replace here */,
};
// do the send here
```